reporting/impl/src/main/java/org/jboss/windup/reporting/export/ExportCSVFileRuleProvider.java

```diff
@@ -183,50 +188,140 @@ private String buildLinkString(Iterable<LinkModel> links)
+        {
+
+
+
+            ApplicationReportService applicationReportService = new ApplicationReportService(event.getGraphContext());
+            TechnologyTagService techTagService = new TechnologyTagService(event.getGraphContext());
+            List<ApplicationReportModel> apps = applicationReportService.findAll();
+            List<TechnologyTagModel> masterTagList = techTagService.findAll().stream().filter(tag -> !tag.getName().equals("Decompiled Java File")).sorted((o1,o2) ->
+
+                            {
+                                String n1 = o1.getName();
+                                String n2 = o2.getName();
+                                return n1.compareToIgnoreCase(n2);
+                            }).collect(Collectors.toList());
+
+            List<String> headerFieldsList = new ArrayList<>();
+            headerFieldsList.add("App Name");
+            ArrayList<String> appNames = new ArrayList<>();
+            ArrayList<ApplicationReportModel> distinctApps = new ArrayList<>();
+
+            apps.forEach(app ->
+            {
+                if(!appNames.contains(app.getProjectModel().getRootFileModel().getFileName())) {
+                    appNames.add(app.getProjectModel().getRootFileModel().getFileName());
+                    distinctApps.add(app);
+                }
+            });
+
+
+
+            masterTagList.forEach( masterTag -> headerFieldsList.add(masterTag==null?"":masterTag.getName()));
+            String[] headerStringsToWrite = headerFieldsList.toArray(new String[0]);
+
+            distinctApps.stream().sorted((o1,o2) ->
+            {
+                String n1 = o1.getProjectModel().getRootFileModel().getFileName();
+                String n2 = o2.getProjectModel().getRootFileModel().getFileName();
+                return n1.compareToIgnoreCase(n2);
+            }).forEachOrdered(app ->
+            {
+                ProjectModelTraversal traversal = new ProjectModelTraversal(app.getProjectModel(),new OnlyOnceTraversalStrategy());
+                ArrayList<TechnologyTagModel> tagsForApp = new ArrayList<>();
+                ArrayList<TechnologyTagModel> structuredTagsForApp = new ArrayList<>();
+
+                ((TreeSet<TechnologyTagModel>)techTagService.findTechnologyTagsForProject(traversal)).stream().filter(tag -> !tag.getName().equals("Decompiled Java File")).sorted((o1,o2) ->
+                {
+                    String n1 = o1.getName();
+                    String n2 = o2.getName();
+                    return n1.compareToIgnoreCase(n2);
+                }).forEachOrdered(tagsForApp::add);
+
+                masterTagList.forEach( masterTag ->
+                {
+                    if (tagsForApp.contains(masterTag))
+                    {
+                        structuredTagsForApp.add(masterTag);
+                    }
+                    else
+                    {
+                        structuredTagsForApp.add(null);
```

Comment:
Assigning a NULL to a List ?

Reply:
the requirement is to create a header row with each tag used in the project listed in alphabetic order. Each application then has their tags listed to match these columns. When an application doesn't contain a particular tag null is added to the application's structured tag list at the index corresponding to that of the tag in the header row so that later tags appear in the correct columns